File: ios/HitCam/Session/StreamSession.swift
import AVFoundation
import Combine
import UIKit

/// Drives one connection to a PC: handshake, pairing, streaming, remote control and reconnects.
/// Published state is updated on the main thread; networking runs on `queue`.
final class StreamSession: ObservableObject {
    enum Phase: Equatable {
        case idle
        case connecting(ServerAddress)
        case pairing(ServerAddress, attemptsLeft: Int, wrongPin: Bool)
        case streaming(ServerAddress, serverName: String)
        case reconnecting(ServerAddress)
        case failed(String)
    }

    @Published private(set) var phase: Phase = .idle
    @Published private(set) var cameraState: CameraState
    @Published private(set) var sentFps: Double = 0
    @Published private(set) var sentKbps: Int = 0

    let camera = CameraController()
    let capabilities = CameraController.capabilities()

    private let queue = DispatchQueue(label: "hitcam.session", qos: .userInitiated)
    private var connection: FramedConnection?
    private var connectionId: UUID?
    private var address: ServerAddress?
    private var encoder: H264Encoder!
    private var timers: [DispatchSourceTimer] = []
    private var reconnectWork: DispatchWorkItem?
    private var lastReceived = Date()
    private var userStopped = true
    private var isStreaming = false
    private var needKeyframe = true
    private var sentToken: String?

    // Owned by `queue`: the session's copy of the camera state and the matching encoder size.
    private var state: CameraState
    private var outputSize: OutputSize

    // Statistics, owned by `queue`.
    private var droppedFrames = 0
    private var framesThisSecond = 0
    private var bytesThisSecond = 0
    private var lastFps: Double = 0
    private var lastKbps = 0
    private var batteryLevel: Double = 0
    private var isCharging = false

    init() {
        let stored = LocalStore.cameraState
        let initial = CameraRules.sanitized(stored, cameraIds: CameraController.availableCameraIds)
        if stored != nil, stored != initial {
            // Saved by an older version or set to something this phone can't do.
            LocalStore.cameraState = initial
        }
        state = initial
        outputSize = CameraRules.outputSize(for: initial)
        cameraState = initial
        encoder = H264Encoder { [weak self] frame in
            self?.queue.async { self?.send(frame) }
        }
        camera.onFrame = { [weak self] sample in
            guard let self, self.isStreamingFlag else { return }
            self.encoder.encode(sample)
        }
        UIDevice.current.isBatteryMonitoringEnabled = true
    }

    // Read from the capture thread; written only on `queue`.
    private var isStreamingFlag: Bool {
        get { streamingLock.withLock { isStreaming } }
        set { streamingLock.withLock { isStreaming = newValue } }
    }
    private let streamingLock = NSLock()

    // MARK: Public API (main thread)

    func connect(to address: ServerAddress) {
        queue.async {
            self.userStopped = false
            self.address = address
            self.openConnection(reconnecting: false)
        }
    }

    func submitPin(_ pin: String) {
        queue.async {
            guard case .pairing = self.currentPhase else { return }
            self.connection?.send(.pairRequest, json: PairRequest(pin: pin))
        }
    }

    func disconnect() {
        queue.async {
            self.userStopped = true
            // Keep the socket out of teardown so Bye is written before it closes.
            let old = self.connection
            self.connection = nil
            self.teardown()
            old?.close(sending: .bye, json: Bye(reason: "user"))
            self.setPhase(.idle)
        }
    }

    /// Local controls from the phone UI go through the same path as controls from the PC.
    func apply(_ control: Control) {
        queue.async { self.handleControl(control) }
    }

    // MARK: Connection lifecycle (queue)

    private func openConnection(reconnecting: Bool) {
        guard let address else { return }
        teardown()
        setPhase(reconnecting ? .reconnecting(address) : .connecting(address))

        // Events from an older, cancelled connection must not tear down this one.
        let id = UUID()
        connectionId = id
        let connection = FramedConnection(address: address, queue: queue) { [weak self] event in
            guard let self, self.connectionId == id else { return }
            self.handle(event)
        }
        self.connection = connection
        connection.start()
    }

    private func handle(_ event: FramedConnection.Event) {
        switch event {
        case .ready:
            lastReceived = Date()
            sendHello()
        case .message(let header, let payload):
            lastReceived = Date()
            handleMessage(header, payload)
        case .closed(let error):
            let wasStreaming = isStreaming
            connectionId = nil
            teardown()
            guard !userStopped, address != nil else { return }
            if wasStreaming, let address {
                // Wi-Fi hiccup or PC restarted: keep trying while the app is open.
                setPhase(.reconnecting(address))
                scheduleReconnect()
            } else if case .reconnecting = currentPhase {
                scheduleReconnect()
            } else if case .failed = currentPhase {
                // Keep the more specific message from the handshake.
            } else {
                fail(error.map { L10n.connectionFailed($0.localizedDescription) } ?? L10n.connectionClosed)
            }
        }
    }

    /// Cancelled by `teardown`, so a new connection (another PC, a PIN being typed) is never replaced by a stale retry.
    private func scheduleReconnect() {
        reconnectWork?.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self, !self.userStopped else { return }
            self.reconnectWork = nil
            self.openConnection(reconnecting: true)
        }
        reconnectWork = work
        queue.asyncAfter(deadline: .now() + 2, execute: work)
    }

    private var currentPhase: Phase = .idle

    private func tokenKeys() -> [String] {
        guard let address else { return [] }
        return [address.serverId, "\(address.host):\(address.port)"].compactMap { $0 }
    }

    private func sendHello() {
        let token = tokenKeys().lazy.compactMap { TokenStore.token(for: $0) }.first
        sentToken = token
        let hello = Hello(
            protocolVersion: ProtocolInfo.version,
            deviceId: LocalStore.deviceId,
            deviceName: UIDevice.current.name,
            model: LocalStore.modelIdentifier,
            appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String,
            token: token)
        connection?.send(.hello, json: hello)
    }

    private func handleMessage(_ header: MessageHeader, _ payload: Data) {
        let decoder = JSONDecoder()
        switch header.type {
        case .helloAck:
            guard let ack = try? decoder.decode(HelloAck.self, from: payload) else { return fail(L10n.protocolError) }
            if let expected = address?.serverId, expected != ack.serverId {
                // The QR code (or an earlier session) named another PC: stop here and send nothing more.
                return fail(L10n.otherPc)
            }
            if address?.serverId == nil { address?.serverId = ack.serverId }
            if address?.name == nil { address?.name = ack.serverName }
            switch ack.status {
            case HelloStatus.accepted:
                startStreaming(serverName: ack.serverName)
            case HelloStatus.pairingRequired:
                forgetRejectedToken()
                if let address { setPhase(.pairing(address, attemptsLeft: 5, wrongPin: false)) }
            case HelloStatus.busy:
                fail(L10n.busy)
            case HelloStatus.pairingLocked:
                fail(L10n.pairingLocked)
            default:
                fail(L10n.versionMismatch)
            }
        case .pairResult:
            guard case .pairing = currentPhase,
                  let result = try? decoder.decode(PairResult.self, from: payload), let address else { return }
            if result.ok, let token = result.token {
                tokenKeys().forEach { TokenStore.save(token, for: $0) }
                startStreaming(serverName: address.name ?? address.host)
            } else if result.attemptsLeft > 0 {
                setPhase(.pairing(address, attemptsLeft: result.attemptsLeft, wrongPin: true))
            } else {
                fail(L10n.pairingLocked)
            }
        case .ping:
            connection?.send(.pong, payload: pongPayload(echo: header.timestamp))
        case .control:
            // Only a paired PC may control the camera.
            guard isStreaming, let control = try? decoder.decode(Control.self, from: payload) else { return }
            handleControl(control)
        case .requestKeyframe:
            guard isStreaming else { return }
            encoder.requestKeyframe()
        case .bye:
            fail(L10n.closedByPc)
        default:
            break
        }
    }

    /// The PC asked for a PIN although a token was sent: it was revoked there, so drop it here too.
    private func forgetRejectedToken() {
        guard let sent = sentToken else { return }
        sentToken = nil
        for key in tokenKeys() where TokenStore.token(for: key) == sent {
            TokenStore.remove(for: key)
        }
    }

    private func pongPayload(echo timestamp: UInt64) -> Data {
        var data = Data()
        data.appendLittleEndian(timestamp)
        return data
    }

    // MARK: Streaming (queue)

    private func startStreaming(serverName: String) {
        guard let address, let id = connectionId else { return }
        LocalStore.remember(address)
        refreshBattery()
        startCamera(with: state, connection: id, address: address, serverName: serverName)
    }

    private func startCamera(with initial: CameraState, connection id: UUID, address: ServerAddress, serverName: String) {
        camera.start(initial: initial) { [weak self] result in
            guard let self else { return }
            self.queue.async {
                // The connection dropped or the user left while the camera was starting.
                // Nothing to undo: teardown queued camera.stop() after this start.
                guard self.connectionId == id, !self.userStopped else { return }
                switch result {
                case .failure(let error):
                    if initial != CameraRules.defaultState {
                        // The saved settings don't work on this phone: start over from the defaults.
                        LocalStore.cameraState = nil
                        self.startCamera(with: CameraRules.defaultState, connection: id, address: address, serverName: serverName)
                    } else {
                        self.fail(L10n.cameraFailed(error.localizedDescription))
                    }
                case .success(let snapshot):
                    self.state = snapshot.state
                    self.outputSize = snapshot.outputSize
                    self.publish(snapshot.state)
                    do {
                        try self.configureEncoder()
                    } catch {
                        return self.fail(L10n.cameraFailed(error.localizedDescription))
                    }
                    self.connection?.send(.capabilities, json: self.capabilities)
                    self.connection?.send(.cameraState, json: snapshot.state)
                    self.isStreamingFlag = true
                    self.startTimers()
                    self.setPhase(.streaming(address, serverName: serverName))
                }
            }
        }
    }

    private func configureEncoder() throws {
        try encoder.configure(width: outputSize.width, height: outputSize.height, fps: state.fps, bitrateKbps: state.bitrateKbps)
        needKeyframe = true
        connection?.send(.streamConfig, json: StreamConfig(
            codec: "h264", width: Int(outputSize.width), height: Int(outputSize.height), fps: state.fps, bitrateKbps: state.bitrateKbps))
    }

    private func send(_ frame: EncodedFrame) {
        guard isStreaming, let connection else { return }

        // Low-latency rule: never let frames queue up. When the socket is congested, drop until a keyframe.
        if connection.framesInFlight >= 2 || (needKeyframe && !frame.isKeyframe) {
            droppedFrames += 1
            if !needKeyframe {
                needKeyframe = true
                encoder.requestKeyframe()
            }
            return
        }
        needKeyframe = false
        connection.send(.videoFrame, flags: frame.isKeyframe ? .keyframe : [], timestamp: frame.timestampMicros,
                        payload: frame.data, isVideo: true)
        framesThisSecond += 1
        bytesThisSecond += frame.data.count
    }

    private func handleControl(_ control: Control) {
        camera.apply(control) { [weak self] update in
            guard let self else { return }
            self.queue.async {
                let state = update.snapshot.state
                self.state = state
                self.outputSize = update.snapshot.outputSize
                self.publish(state)
                // A rejected format was rolled back; only a state the camera accepted is remembered.
                if !update.failed, CameraRules.isValid(state, cameraIds: CameraController.availableCameraIds) {
                    LocalStore.cameraState = state
                }
                guard self.isStreaming else { return }
                if update.formatChanged {
                    do {
                        try self.configureEncoder()
                    } catch {
                        return self.fail(L10n.cameraFailed(error.localizedDescription))
                    }
                } else if update.bitrateChanged {
                    // The PC reads the new bitrate from the camera state below; no new StreamConfig needed.
                    self.encoder.setBitrate(kbps: state.bitrateKbps)
                }
                self.connection?.send(.cameraState, json: state)
            }
        }
    }

    private func startTimers() {
        stopTimers()
        timers.append(repeating(1) { [weak self] in
            guard let self else { return }
            self.connection?.send(.ping)
            let fps = Double(self.framesThisSecond)
            let kbps = self.bytesThisSecond * 8 / 1000
            self.framesThisSecond = 0
            self.bytesThisSecond = 0
            self.lastFps = fps
            self.lastKbps = kbps
            DispatchQueue.main.async {
                self.sentFps = fps
                self.sentKbps = kbps
            }
            if Date().timeIntervalSince(self.lastReceived) > 5 {
                self.connection?.cancel()   // liveness timeout -> reconnect path
            }
        })
        timers.append(repeating(2) { [weak self] in
            self?.sendStatus()
        })
    }

    /// UIDevice is read on the main thread; the values reach `queue` a moment later.
    private func refreshBattery() {
        DispatchQueue.main.async { [weak self] in
            let device = UIDevice.current
            let level = Double(max(device.batteryLevel, 0))
            let charging = device.batteryState == .charging || device.batteryState == .full
            guard let self else { return }
            self.queue.async {
                self.batteryLevel = level
                self.isCharging = charging
            }
        }
    }

    private func sendStatus() {
        let thermal: String
        switch ProcessInfo.processInfo.thermalState {
        case .nominal: thermal = "nominal"
        case .fair: thermal = "fair"
        case .serious: thermal = "serious"
        case .critical: thermal = "critical"
        @unknown default: thermal = "nominal"
        }
        let status = DeviceStatus(
            battery: batteryLevel, charging: isCharging,
            thermal: thermal, fps: lastFps, bitrateKbps: lastKbps, droppedFrames: droppedFrames)
        connection?.send(.status, json: status)
        refreshBattery()
    }

    private func repeating(_ seconds: Double, _ body: @escaping () -> Void) -> DispatchSourceTimer {
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + seconds, repeating: seconds)
        timer.setEventHandler(handler: body)
        timer.resume()
        return timer
    }

    private func stopTimers() {
        timers.forEach { $0.cancel() }
        timers.removeAll()
    }

    private func teardown() {
        isStreamingFlag = false
        reconnectWork?.cancel()
        reconnectWork = nil
        stopTimers()
        encoder.invalidate()
        camera.stop()
        let old = connection
        connection = nil
        connectionId = nil
        old?.cancel()
    }

    // MARK: State publishing

    private func fail(_ message: String) {
        userStopped = true
        teardown()
        setPhase(.failed(message))
    }

    private func setPhase(_ phase: Phase) {
        currentPhase = phase
        DispatchQueue.main.async {
            self.phase = phase
            UIApplication.shared.isIdleTimerDisabled = {
                if case .streaming = phase { return true }
                if case .reconnecting = phase { return true }
                return false
            }()
        }
    }

    private func publish(_ state: CameraState) {
        DispatchQueue.main.async { self.cameraState = state }
    }
}

File: android/app/src/main/java/io/github/hitnes/hitcam/session/StreamSession.kt
package io.github.hitnes.hitcam.session

import io.github.hitnes.hitcam.camera.CameraRules
import io.github.hitnes.hitcam.camera.OutputSize
import io.github.hitnes.hitcam.net.FramedConnection
import io.github.hitnes.hitcam.protocol.Bye
import io.github.hitnes.hitcam.protocol.CameraState
import io.github.hitnes.hitcam.protocol.Capabilities
import io.github.hitnes.hitcam.protocol.Control
import io.github.hitnes.hitcam.protocol.DeviceStatus
import io.github.hitnes.hitcam.protocol.Hello
import io.github.hitnes.hitcam.protocol.HelloAck
import io.github.hitnes.hitcam.protocol.HelloStatus
import io.github.hitnes.hitcam.protocol.MessageFlags
import io.github.hitnes.hitcam.protocol.MessageHeader
import io.github.hitnes.hitcam.protocol.MessageType
import io.github.hitnes.hitcam.protocol.PairRequest
import io.github.hitnes.hitcam.protocol.PairResult
import io.github.hitnes.hitcam.protocol.ProtocolInfo
import io.github.hitnes.hitcam.protocol.ProtocolJson
import io.github.hitnes.hitcam.protocol.ServerAddress
import io.github.hitnes.hitcam.protocol.StreamConfig
import io.github.hitnes.hitcam.protocol.pongPayload
import io.github.hitnes.hitcam.video.EncodedFrame
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/** Why a session ended; the UI turns it into text. */
sealed interface SessionError {
    data class ConnectionFailed(val reason: String) : SessionError
    data object ConnectionClosed : SessionError
    data object ProtocolError : SessionError
    data object OtherPc : SessionError
    data object Busy : SessionError
    data object PairingLocked : SessionError
    data object VersionMismatch : SessionError
    data object ClosedByPc : SessionError
    data class CameraFailed(val reason: String) : SessionError
}

sealed interface Phase {
    data object Idle : Phase
    data class Connecting(val address: ServerAddress) : Phase
    data class Pairing(val address: ServerAddress, val attemptsLeft: Int, val wrongPin: Boolean) : Phase
    data class Streaming(val address: ServerAddress, val serverName: String) : Phase
    data class Reconnecting(val address: ServerAddress) : Phase
    data class Failed(val error: SessionError) : Phase
}

/**
 * Drives one connection to a PC: handshake, pairing, streaming, remote control and reconnects.
 * All state lives on one thread ([queue]); the UI observes [phase], [cameraState] and the send statistics.
 */
class StreamSession(
    private val video: VideoSource,
    private val environment: SessionEnvironment,
    private val reconnectDelayMs: Long = 2_000,
) {
    private val queue = Executors.newSingleThreadScheduledExecutor { Thread(it, "hitcam-session") }

    private val _phase = MutableStateFlow<Phase>(Phase.Idle)
    val phase: StateFlow<Phase> = _phase.asStateFlow()
    private val _cameraState: MutableStateFlow<CameraState>
    val cameraState: StateFlow<CameraState>
    private val _sentFps = MutableStateFlow(0.0)
    val sentFps: StateFlow<Double> = _sentFps.asStateFlow()
    private val _sentKbps = MutableStateFlow(0)
    val sentKbps: StateFlow<Int> = _sentKbps.asStateFlow()

    val capabilities: Capabilities get() = video.capabilities

    // Owned by `queue`.
    private var connection: FramedConnection? = null
    private var connectionId = 0
    private var address: ServerAddress? = null
    private var timers = mutableListOf<ScheduledFuture<*>>()
    private var reconnect: ScheduledFuture<*>? = null
    private var lastReceived = 0L
    private var userStopped = true
    private var cameraPaused = false
    private var needKeyframe = true
    private var sentToken: String? = null
    private var serverName = ""
    private var state: CameraState
    private var outputSize: OutputSize

    // Read from the encoder thread; written only on `queue`.
    @Volatile private var isStreaming = false

    // Statistics, owned by `queue`.
    private var droppedFrames = 0
    private var framesThisSecond = 0
    private var bytesThisSecond = 0
    private var lastFps = 0.0
    private var lastKbps = 0

    init {
        val stored = environment.cameraState
        val initial = CameraRules.sanitized(stored, video.cameraIds)
        // Saved by an older version or set to something this phone can't do.
        if (stored != null && stored != initial) environment.cameraState = initial
        state = initial
        outputSize = CameraRules.outputSize(initial)
        _cameraState = MutableStateFlow(initial)
        cameraState = _cameraState.asStateFlow()
        video.onFrame = { frame -> if (isStreaming) queue.execute { send(frame) } }
        video.onError = { message -> queue.execute { if (isStreaming) fail(SessionError.CameraFailed(message)) } }
    }

    // MARK: Public API (any thread)

    fun connect(address: ServerAddress) = queue.execute {
        userStopped = false
        this.address = address
        openConnection(reconnecting = false)
    }

    fun submitPin(pin: String) = queue.execute {
        if (_phase.value is Phase.Pairing) connection?.send(MessageType.PairRequest, PairRequest.serializer(), PairRequest(pin))
    }

    fun disconnect() = queue.execute {
        userStopped = true
        // Keep the socket out of teardown so Bye is written before it closes.
        val old = connection
        connection = null
        teardown()
        old?.close(MessageType.Bye, Bye.serializer(), Bye("user"))
        setPhase(Phase.Idle)
    }

    /** Local controls from the phone UI go through the same path as controls from the PC. */
    fun apply(control: Control) = queue.execute { handleControl(control) }

    /** The app went to the background: Android takes the camera away, so release it but keep the connection. */
    fun pauseCamera() = queue.execute {
        if (cameraPaused) return@execute
        cameraPaused = true
        if (isStreaming) {
            isStreaming = false
            video.stop()
        }
    }

    fun resumeCamera() = queue.execute {
        if (!cameraPaused) return@execute
        cameraPaused = false
        val phase = _phase.value
        if (phase is Phase.Streaming && connection != null) startCamera(state, connectionId, phase.address, phase.serverName, resumed = true)
    }

    fun close() {
        disconnect()
        queue.shutdown()
    }

    // MARK: Connection lifecycle (queue)

    private fun openConnection(reconnecting: Boolean) {
        val address = address ?: return
        teardown()
        setPhase(if (reconnecting) Phase.Reconnecting(address) else Phase.Connecting(address))

        // Events from an older, cancelled connection must not tear down this one.
        val id = ++connectionId
        val connection = FramedConnection(address.host, address.port, queue) { event ->
            if (connectionId == id) handle(event)
        }
        this.connection = connection
        connection.start()
    }

    private fun handle(event: FramedConnection.Event) {
        when (event) {
            is FramedConnection.Event.Ready -> {
                lastReceived = System.nanoTime()
                sendHello()
            }
            is FramedConnection.Event.Message -> {
                lastReceived = System.nanoTime()
                handleMessage(event.header, event.payload)
            }
            is FramedConnection.Event.Closed -> {
                val wasStreaming = _phase.value is Phase.Streaming
                connectionId++
                teardown()
                val address = address
                if (userStopped || address == null) return
                val phase = _phase.value
                when {
                    wasStreaming -> {
                        // Wi-Fi hiccup or PC restarted: keep trying while the app is open.
                        setPhase(Phase.Reconnecting(address))
                        scheduleReconnect()
                    }
                    phase is Phase.Reconnecting -> scheduleReconnect()
                    phase is Phase.Failed -> Unit // Keep the more specific message from the handshake.
                    else -> fail(event.error?.let { SessionError.ConnectionFailed(it.message ?: it.javaClass.simpleName) } ?: SessionError.ConnectionClosed)
                }
            }
        }
    }

    /** Cancelled by [teardown], so a new connection is never replaced by a stale retry. */
    private fun scheduleReconnect() {
        reconnect?.cancel(false)
        reconnect = queue.schedule({
            reconnect = null
            if (!userStopped) openConnection(reconnecting = true)
        }, reconnectDelayMs, TimeUnit.MILLISECONDS)
    }

    private fun tokenKeys(): List<String> {
        val address = address ?: return emptyList()
        return listOfNotNull(address.serverId, "${address.host}:${address.port}")
    }

    private fun sendHello() {
        val token = tokenKeys().firstNotNullOfOrNull { environment.token(it) }
        sentToken = token
        val hello = Hello(
            protocolVersion = ProtocolInfo.VERSION,
            deviceId = environment.deviceId,
            deviceName = environment.deviceName,
            model = environment.model,
            appVersion = environment.appVersion,
            token = token,
        )
        connection?.send(MessageType.Hello, Hello.serializer(), hello)
    }

    private fun handleMessage(header: MessageHeader, payload: ByteArray) {
        when (header.type) {
            MessageType.HelloAck -> {
                val ack = decode<HelloAck>(payload) ?: return fail(SessionError.ProtocolError)
                val expected = address?.serverId
                // The QR code (or an earlier session) named another PC: stop here and send nothing more.
                if (expected != null && expected != ack.serverId) return fail(SessionError.OtherPc)
                address = address?.let {
                    it.copy(serverId = it.serverId ?: ack.serverId, name = it.name ?: ack.serverName)
                }
                when (ack.status) {
                    HelloStatus.ACCEPTED -> startStreaming(ack.serverName)
                    HelloStatus.PAIRING_REQUIRED -> {
                        forgetRejectedToken()
                        address?.let { setPhase(Phase.Pairing(it, attemptsLeft = 5, wrongPin = false)) }
                    }
                    HelloStatus.BUSY -> fail(SessionError.Busy)
                    HelloStatus.PAIRING_LOCKED -> fail(SessionError.PairingLocked)
                    else -> fail(SessionError.VersionMismatch)
                }
            }
            MessageType.PairResult -> {
                if (_phase.value !is Phase.Pairing) return
                val result = decode<PairResult>(payload) ?: return
                val address = address ?: return
                val token = result.token
                when {
                    result.ok && token != null -> {
                        tokenKeys().forEach { environment.saveToken(token, it) }
                        startStreaming(address.name ?: address.host)
                    }
                    result.attemptsLeft > 0 -> setPhase(Phase.Pairing(address, result.attemptsLeft, wrongPin = true))
                    else -> fail(SessionError.PairingLocked)
                }
            }
            MessageType.Ping -> connection?.send(MessageType.Pong, pongPayload(header.timestamp))
            MessageType.Control -> {
                // Only a paired PC may control the camera.
                if (_phase.value !is Phase.Streaming) return
                decode<Control>(payload)?.let { handleControl(it) }
            }
            MessageType.RequestKeyframe -> if (isStreaming) video.requestKeyframe()
            MessageType.Bye -> fail(SessionError.ClosedByPc)
            else -> Unit
        }
    }

    private inline fun <reified T> decode(payload: ByteArray): T? =
        runCatching { ProtocolJson.decodeFromString<T>(payload.decodeToString()) }.getOrNull()

    /** The PC asked for a PIN although a token was sent: it was revoked there, so drop it here too. */
    private fun forgetRejectedToken() {
        val sent = sentToken ?: return
        sentToken = null
        tokenKeys().filter { environment.token(it) == sent }.forEach { environment.removeToken(it) }
    }

    // MARK: Streaming (queue)

    private fun startStreaming(serverName: String) {
        val address = address ?: return
        this.serverName = serverName
        environment.remember(address)
        setPhase(Phase.Streaming(address, serverName))
        startTimers()
        if (!cameraPaused) startCamera(state, connectionId, address, serverName, resumed = false)
    }

    private fun startCamera(initial: CameraState, id: Int, address: ServerAddress, serverName: String, resumed: Boolean) {
        video.start(initial) { result ->
            queue.execute {
                // The connection dropped, the user left or the app went to the background while the camera was starting.
                if (connectionId != id || userStopped || cameraPaused) {
                    if (connectionId == id && cameraPaused) video.stop()
                    return@execute
                }
                result.onFailure { error ->
                    if (initial != CameraRules.defaultState) {
                        // The saved settings don't work on this phone: start over from the defaults.
                        environment.cameraState = null
                        startCamera(CameraRules.defaultState, id, address, serverName, resumed)
                    } else {
                        fail(SessionError.CameraFailed(error.message ?: error.javaClass.simpleName))
                    }
                }
                result.onSuccess { snapshot ->
                    state = snapshot.state
                    outputSize = snapshot.outputSize
                    _cameraState.value = snapshot.state
                    try {
                        configureEncoder()
                    } catch (e: Exception) {
                        return@execute fail(SessionError.CameraFailed(e.message ?: e.javaClass.simpleName))
                    }
                    if (!resumed) connection?.send(MessageType.Capabilities, Capabilities.serializer(), video.capabilities)
                    connection?.send(MessageType.CameraState, CameraState.serializer(), snapshot.state)
                    isStreaming = true
                }
            }
        }
    }

    private fun configureEncoder() {
        video.configureEncoder(outputSize, state.fps, state.bitrateKbps)
        needKeyframe = true
        connection?.send(
            MessageType.StreamConfig, StreamConfig.serializer(),
            StreamConfig("h264", outputSize.width, outputSize.height, state.fps, state.bitrateKbps),
        )
    }

    private fun send(frame: EncodedFrame) {
        val connection = connection
        if (!isStreaming || connection == null) return

        // Low-latency rule: never let frames queue up. When the socket is congested, drop until a keyframe.
        if (connection.framesInFlight >= 2 || (needKeyframe && !frame.isKeyframe)) {
            droppedFrames++
            if (!needKeyframe) {
                needKeyframe = true
                video.requestKeyframe()
            } else if (!frame.isKeyframe) {
                video.requestKeyframe()
            }
            return
        }
        needKeyframe = false
        connection.send(
            MessageType.VideoFrame, frame.data, flags = if (frame.isKeyframe) MessageFlags.KEYFRAME else 0,
            timestamp = frame.timestampMicros, isVideo = true,
        )
        framesThisSecond++
        bytesThisSecond += frame.data.size
    }

    private fun handleControl(control: Control) {
        video.apply(control) { update ->
            queue.execute {
                val state = update.snapshot.state
                this.state = state
                outputSize = update.snapshot.outputSize
                _cameraState.value = state
                // A rejected format was rolled back; only a state the camera accepted is remembered.
                if (!update.failed && CameraRules.isValid(state, video.cameraIds)) environment.cameraState = state
                if (!isStreaming) return@execute
                if (update.formatChanged) {
                    try {
                        configureEncoder()
                    } catch (e: Exception) {
                        return@execute fail(SessionError.CameraFailed(e.message ?: e.javaClass.simpleName))
                    }
                } else if (update.bitrateChanged) {
                    // The PC reads the new bitrate from the camera state below; no new StreamConfig needed.
                    video.setBitrate(state.bitrateKbps)
                }
                connection?.send(MessageType.CameraState, CameraState.serializer(), state)
            }
        }
    }

    private fun startTimers() {
        stopTimers()
        timers += queue.scheduleWithFixedDelay({ everySecond() }, 1, 1, TimeUnit.SECONDS)
        timers += queue.scheduleWithFixedDelay({ sendStatus() }, 2, 2, TimeUnit.SECONDS)
    }

    private fun everySecond() {
        connection?.send(MessageType.Ping)
        lastFps = framesThisSecond.toDouble()
        lastKbps = bytesThisSecond * 8 / 1000
        framesThisSecond = 0
        bytesThisSecond = 0
        _sentFps.value = lastFps
        _sentKbps.value = lastKbps
        if (System.nanoTime() - lastReceived > TimeUnit.SECONDS.toNanos(5)) {
            // Liveness timeout: handled like a dropped connection (reconnect while streaming).
            val connection = connection ?: return
            this.connection = null
            connection.cancel()
            handle(FramedConnection.Event.Closed(java.io.IOException("timeout")))
        }
    }

    private fun sendStatus() {
        if (_phase.value !is Phase.Streaming) return
        val (battery, charging) = environment.battery()
        val status = DeviceStatus(battery, charging, environment.thermal(), lastFps, lastKbps, droppedFrames)
        connection?.send(MessageType.Status, DeviceStatus.serializer(), status)
    }

    private fun stopTimers() {
        timers.forEach { it.cancel(false) }
        timers.clear()
    }

    private fun teardown() {
        isStreaming = false
        reconnect?.cancel(false)
        reconnect = null
        stopTimers()
        video.stop()
        val old = connection
        connection = null
        old?.cancel()
        _sentFps.value = 0.0
        _sentKbps.value = 0
    }

    private fun fail(error: SessionError) {
        userStopped = true
        teardown()
        setPhase(Phase.Failed(error))
    }

    private fun setPhase(phase: Phase) {
        _phase.value = phase
    }
}
